engine/enqueue.go

```diff
@@ -26,16 +26,18 @@ type Pattern struct {
 	Match MatchFunc
 	// Map is a function to apply to messages before storing them. If not provided, then the message is stored in its original form.
 	Map MapFunc
+	// Validate is a function to validate a message before storing it. If not provided, then the message is always stored.
+	Validate ValidateFunc
```

Comment:
We're solving the same problem slightly different ways. Here's a PR on the Pacemaker branch addressing a similar problem: https://github.com/onflow/flow-go/pull/3470

I'll review the rest of this PR in isolation, but we should align on one of these approaches.